src/client/Client.cc

```diff
@@ -5421,6 +5421,12 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   int r = _getattr_for_perm(in, perms);
   if (r < 0)
     goto out;
+  
+  if (in->mode & S_ISUID)
+    goto out;
+
+  if ((in->mode & (S_ISGID | S_IXGRP)) == (S_ISGID | S_IXGRP))
+    goto out;
```

Comment:
this allow all operation for file with setuid/setgid, which is incorrect.
In may_setattr CEPH_SETATTR_MODE backet, we should allow removing SUID/SGID if file write is allowed for uid.
A better fix is set FUSE_HANDLE_KILLPRIV fuse flag, and let libcephfs handle removing suid/sgid on write.

Reply:
something like:
<pre>
diff --git a/src/client/Client.cc b/src/client/Client.cc
index 84fcf91d0d..25a917c704 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -5394,8 +5394,16 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (mask & CEPH_SETATTR_MODE) {
-    if (perms.uid() != 0 && perms.uid() != in->uid)
-      goto out;
+    if (perms.uid() != 0 && perms.uid() != in->uid) {
+      auto to_clear = ~stx->stx_mode & in->mode & 07777;
+      if (to_clear && !(to_clear & ~(S_ISUID | S_ISGID)) && // only remove setuid/setgid
+         !(stx->stx_mode & ~in->mode & 07777) && // not add new bits
+         !inode_permission(in, perms, MAY_WRITE)) { // may write
+       // hack: kill setuid/setguid on write?
+      } else {
+       goto out;
+      }
+    }
 
     gid_t i_gid = (mask & CEPH_SETATTR_GID) ? stx->stx_gid : in->gid;
     if (perms.uid() != 0 && !perms.gid_in_groups(i_gid))

</pre>

Reply:
Thank you
FUSE_HANDLE_KILLPRIV fuse flag is the best way , but only supported in fuse3.0

Reply:
！I first try to test it

Reply:
The provided method cannot be verified, and the write function will clear out suid and sgid